src/ITOps.ViewModelComposition/DynamicViewModel.cs

```diff
@@ -0,0 +1,93 @@
+﻿namespace ITOps.ViewModelComposition
+{
+    using Microsoft.AspNetCore.Http;
+    using Microsoft.AspNetCore.Routing;
+    using System;
+    using System.Collections.Generic;
+    using System.Dynamic;
+    using System.Threading.Tasks;
+
+    public class DynamicViewModel : DynamicObject
+    {
+        // Keep track of the routeData and the query collection that's being passed in.
+        private readonly RouteData routeData;
+        private readonly IQueryCollection query;
+        
+        // To extend and keep track of properties as part of the dynamic object.
+        private IDictionary<string, object> properties = new Dictionary<string, object>();
+
+        // Keep a list of subcribers of the same route interested in getting called when events are raised.
+        private IDictionary<Type, List<EventHandler<object>>> callbackRegistrations = new Dictionary<Type, List<EventHandler<object>>>();
+
+        // Needed for registering parameterized callbacks from subscribers. 
+        public delegate Task EventHandler<TEvent>(dynamic pageViewModel, TEvent @event, RouteData routeData, IQueryCollection query);
```

Comment:
Nitpicking: I’d move this to a separate file as it is effectively a type declaration.

Reply:
The only reason I wanted it in the same file, was it made it so much easier to understand, because its part of the class.